### DISCUSSION TITLE
bounding box, label, confidence score missing.

### DISCUSSION THREAD

**Discussion by aiyafi** [ultralytics/ultralytics#10348] [General]
I need help, I want to make a program about “Loitering Detection in an Intelligent Surveillance System using YOLO-based Methods”. Where I have made it as in the following picture:
![image](https://github.com/ultralytics/ultralytics/assets/63116789/60b1e72f-cd98-4370-aa9f-5d08a1440731)

Then on the centroid feature I have created and it works. But I want the bounding box, label, and confidence score to remain. Then how do I make it exist?

Just look like this repository:
[https://github.com/ultralytics/yolov5/issues/6636](url)

I want to run the code, but there is an error and this is still using Yolo version 5, while in my project using Yolo version 8n. So please, can someone help me out?

and here's the full code:

```
from ultralytics import YOLO, engine
import cv2

# Model setup
#confidence_threshold = 0.8
model = YOLO('yolov8n.pt')
result = model.predict(source="./PGM.mp4", save=True, save_txt=True, classes=[0], stream=True)  # save=True saves the results to a folder

# Video loading
video_path = './PGM-normal_cewek_sore_2.mp4'
cap = cv2.VideoCapture(video_path)

# Get video dimensions
width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))

# Main processing loop
while True:
    ret, frame = cap.read()
    if not ret:
        break  # End of video

    # Optional resizing
    frame = cv2.resize(frame, (960, 540)) 

    # Object detection and centroid calculation
    results = model.predict(frame, save=True, save_txt=True, classes=[0])  

    for r in results:  
        for box in r.boxes.xyxy:
            centroid_x = (box[0] + box[2]) / 2
            centroid_y = (box[1] + box[3]) / 2

            # Print and/or draw on the frame
            print(f"Centroid Coordinates: ({centroid_x}, {centroid_y})")
            cv2.circle(frame, (int(centroid_x), int(centroid_y)), radius=2, color=(0, 0, 255), thickness=3) # Draw centroid

    # Display the frame with bounding boxes and centroids
    cv2.imshow('frame', frame) 

    # Exit condition
    if cv2.waitKey(25) & 0xFF == ord('q'):
        break

# Release resources
cap.release()
cv2.destroyAllWindows()
```

**Comment by glenn-jocher** [comment]
Hello 👋, thanks for reaching out with your question!

From your description, it seems like you want to retain the bounding box, label, and confidence score information on your video frames. It appears there may have been a confusion in implementing the drawing and labeling of bounding boxes and scores within your loop.

Here's a quick example snippet illustrating how you could enhance your code to display bounding boxes, labels, and confidence scores:

```python
# ... Your existing setup code ...

while True:
    ret, frame = cap.read()
    if not ret:
        break  # End of video
    
    # ... Your existing resizing and prediction code ...

    for r in results:  
        for i, box in enumerate(r.boxes.xyxy):
            centroid_x = (box[0] + box[2]) / 2
            centroid_y = (box[1] + box[3]) / 2

            # Modify here: Draw bounding box and add label + confidence score
            label = f'{r.names[r.boxes.cls[i]]} {r.boxes.conf[i]:.2f}'
            cv2.rectangle(frame, (int(box[0]), int(box[1])), (int(box[2]), int(box[3])), (0, 255, 0), 2)
            cv2.putText(frame, label, (int(box[0]), int(box[1]-10)), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (0,255,0), 2)

            # Print and/or draw on the frame
            print(f"Centroid Coordinates: ({centroid_x}, {centroid_y})")
            cv2.circle(frame, (int(centroid_x), int(centroid_y)), radius=2, color=(0, 0, 255), thickness=3) # Draw centroid

    # ... Your existing display and exit code ...
```

This snippet specifically adds rectangles and text to your frames by making use of `cv2.rectangle` and `cv2.putText` methods respectively, where:
- `r.names[r.boxes.cls[i]]` retrieves the label for the detected object.
- `r.boxes.conf[i]` retrieves the confidence score.

Ensure to replace the `[0]` in `r.boxes.cls[i]` and `r.boxes.conf[i]` with appropriate indices if your `results` object structure requires it.

Lastly, if you're encountering a specific error with the code or conversion between YOLOv5 and YOLOv8 models, it would be handy to know the exact error message you're encountering for a more targeted solution!

**Reply by aiyafi** [reply]
Hey, thx for the help. Initially, the code runs, but when a human is present (object detection), then the frame stops and issues an error like the following:

```0: 384x640 1 person, 55.4ms
Speed: 1.0ms preprocess, 55.4ms inference, 0.0ms postprocess per image at shape (1, 3, 384, 640)
Results saved to runs\detect\predict86
1 label saved to runs\detect\predict86\labels
Traceback (most recent call last):
  File "D:\@Dev\finalPrj\objectDetection1\5main.py", line 35, in <module>
    label = f'{r.names[r.boxes.cls[i]]} {r.boxes.conf[i]:.2f}'
               ~~~~~~~^^^^^^^^^^^^^^^^
KeyError: tensor(0.)
```

So what's wrong with my code?

**Reply by mustani** [reply]
you can try with "r.names[r.boxes.cls[0]]".

**Reply by glenn-jocher** [reply]
Hey there! 😊 It looks like you're trying to access class names with tensor indices directly. When accessing elements in `r.boxes.cls`, which are tensors, you should first convert them to integer indices like this:

```python
label = f'{r.names[int(r.boxes.cls[i])]} {r.boxes.conf[i].item():.2f}'
```

Using `.item()` with `r.boxes.conf[i]` ensures you're working with Python scalar values rather than PyTorch tensors. This should help avoid the `KeyError` you're encountering. Happy coding! 🚀
